### ISSUE TITLE
softmax in tensorflow and lasagne

### ISSUE THREAD

**Issue by saitarslanboun** [Lasagne/Lasagne#885]
Hi,

In tensorflow, a 5D tensor is given to tensorflow.nn.softmax(input, dim=2) function.

I would like to do the same in lasagne or theano, and I am very new in theano. Could you help me to create an equivalent function in theano or lasagne?

Thanks,

**Comment by rafaelvalle** [comment]
I don't know how theano or lasagne operates right now but suggest you to
reshape your 5d data into 2d(-1, n_categories), apply softmax and then
reshape back to 5d

**Comment by saitarslanboun** [comment]
Thank you Rafael, what does "n_categories" represent?

**Comment by rafaelvalle** [comment]
The size of the dimension whose activations are being normalized with
softmax. In video data, eg, your input can be 5d (batch size, time, x, y,
n_category), where each pixel had a distribution over categories

**Comment by f0k** [comment]
I also started implementing a wrapper class to allow specifying an arbitrary set of axes for the softmax: https://github.com/f0k/Lasagne/commit/807763f64bbdcb8dfbe3ef415c030495ff34d4e8#diff-e803e2c6fef365217ea4b830787511e0R63-144
You can try this (just copy the class into your own code), but it's untested.

**Comment by saitarslanboun** [comment]
Let's say, in tensorflow, I have the following line:

tf.nn.softmax(b_IJ, dim=2)

b_IJ has the shape [ 128 1152   10    1    1].

You gave me the implementation, on this link: https://github.com/f0k/Lasagne/commit/807763f64bbdcb8dfbe3ef415c030495ff34d4e8#diff-e803e2c6fef365217ea4b830787511e0R63-144.

The issue is, this implementation is layer-wise. I need an element-wise implementation. Is it possible, or could you help me or show me, how I can rewrite it in theano?

I would be so thankful.

Kind Regards,
